specifyweb/frontend/js_src/lib/components/SchemaConfig/UniquenessRules.tsx

```diff
@@ -0,0 +1,431 @@
+import React from 'react';
+import { useParams } from 'react-router-dom';
+import type { LocalizedString } from 'typesafe-i18n';
+
+import { commonText } from '../../localization/common';
+import { schemaText } from '../../localization/schema';
+import { ajax } from '../../utils/ajax';
+import type { RA } from '../../utils/types';
+import { defined, filterArray, localized } from '../../utils/types';
+import {
+  insertItem,
+  removeItem,
+  removeKey,
+  replaceItem,
+  sortFunction,
+  split,
+} from '../../utils/utils';
+import { Button } from '../Atoms/Button';
+import { className } from '../Atoms/className';
+import { icons } from '../Atoms/Icons';
+import { Submit } from '../Atoms/Submit';
+import { LoadingContext } from '../Core/Contexts';
+import { addMissingFields } from '../DataModel/addMissingFields';
+import type { SerializedResource } from '../DataModel/helperTypes';
+import { schema } from '../DataModel/schema';
+import type { Relationship, RelationshipType } from '../DataModel/specifyField';
+import { getTable, strictGetTable } from '../DataModel/tables';
+import type {
+  SpLocaleContainer,
+  SpLocaleContainerItem,
+  Tables,
+} from '../DataModel/types';
+import type {
+  UniquenessRule,
+  UniquenessRuleValidation,
+} from '../DataModel/uniquenessRules';
+import {
+  databaseFieldName,
+  databaseScope,
+  getUniqueInvalidReason,
+  useTableUniquenessRules,
+  validateUniqueness,
+} from '../DataModel/uniquenessRules';
+import { raise } from '../Errors/Crash';
+import { Dialog } from '../Molecules/Dialog';
+import { userPreferences } from '../Preferences/userPreferences';
+import type { WithFetchedStrings } from '../Toolbar/SchemaConfig';
+import { downloadDataSet } from '../WorkBench/helpers';
+import { PickList } from './Components';
+import { useContainerItems } from './Hooks';
+
+export function TableUniquenessRules({
+  container,
+  header,
+  onClose: handleClose,
+}: {
+  readonly container: SerializedResource<SpLocaleContainer>;
+  readonly header: LocalizedString;
+  readonly onClose: () => void;
+}): JSX.Element {
+  const { language: rawLanguage = '' } = useParams();
+  const [language, country = null] = rawLanguage.split('-');
+
+  const [tableRules = [], setTableRules, setCachedTableRules] =
+    useTableUniquenessRules(strictGetTable(container.name).name);
+
+  const loading = React.useContext(LoadingContext);
+
+  const [saveBlocked, setSavedBlocked] = React.useState(false);
+
+  const [isRuleExpanded, setRuleExpanded] = React.useState<
+    Record<number, boolean>
+  >({});
+
+  const [allItems = []] = useContainerItems(container, language, country);
+
+  const allFieldNames = React.useMemo(
+    () => Object.values(allItems).sort(sortFunction(({ name }) => name)),
+    [allItems]
+  );
+
+  const [fields = [], relationships = []] = React.useMemo(() => {
```

Comment:
Do we need `fields = [], relationships = []`? It could simply be `fields, relationships` (no need for `=[]`)

Reply:
Resolved in https://github.com/specify/specify7/pull/4215/commits/b15d6ab6d89fa8a1e5fa9101a1760465cddf5d7b. 
I believe there was a time when `fields` and `relationships` could be undefined, and thus the default values were needed. 
Not the case anymore. Thanks for catching that!